### PR TITLE
Completion Handlers, iCloud Status Monitoring, Per-User Local Cache Changes FOR REVIEW

### PR DESCRIPTION
**Pull request by tbaggett**
Hi Edwin,

These changes ended up being pretty substantial. As I mentioned before, my experience with iOS/tvOS is limited, and my app, along with your project, is the first Swift coding I've done. I have no doubt that these changes could be implemented differently, and could probably be implemented in better, simpler, and/or smarter ways. I welcome your review and comments. I have already learned a lot from your project, and fully expect to learn more from us discussing these changes. I hope your schedule will allow you to do so.

Here's some notes on my thinking behind the changes:

- IMHO, adding the completion handlers while maintaining the publicDB, privateDB and ...forContainer DB constants required keeping the two fully separate. I also thought that being able to assign multiple completion handlers was a reasonable expectation. My changes support that behavior.

- I also included support for automatically calling completion handlers that are added after a constant DB entity is initialized. The new handlers are called the next time the constant entity is referenced. All assigned handlers are called again in the event of a DB reinitialization.

- A wrapper class was needed to facilitate easy removal of handlers if they were no longer needed. The wrapper class is returned as a token that the calling code is expected to keep up with, then call the wrapper's releaseToken method when the associated handler should no longer be called. I also created a manager class that I use in my app to keep up with the tokens and remove all of them when it goes out of scope. I can include that if you'd like.

- Per-user caching now uses the iCloud account's user recordId as part of the file name. This change required some semaphore usage to wait on the recordId to be retrieved the first time. Also, the local cache file name is constructed the same as before if no user recordId is returned. This may need to be changed.

- I had to change the iCloud status monitoring code from the sample code I originally provided in the feature request. I found that code was broken in varying ways depending on the OS version being used. It now uses fetchUserRecordIDWithCompletionHandler, which is also used to construct the local cache file names.

- Currently, an EVCloudKitDao DB instance has its activeUser and (new) activeUserId properties set to nil, followed by the initializeDataBase method being called, for all public and private DB instances when an iCloud status change is detected. I'm not sure if anything else needs to be reset or not (subscriptions maybe?). Please let me know or make the needed changes if other things are needed.

- EVCloudData adds a completion handler that calls disconnectAll() to each EVCloudKitDao instance that its constant entities depend on. This completion handler is given a higher priority than app-provided ones. It should always be called first, avoiding the possibility of disconnecting everything after the app-provided handlers added their needed connections.

- Being a Swift newb, I am still a little fuzzy on when closures can cause a circular reference that prevents them and their enclosing code from being released from memory. I've read about things like specifying [weak self] and [unowned self] in closures. However I'm hesitant to apply them since they can also cause premature releasing of things before everything finishes executing. I will eventually get around to using the profiler and getting a clearer understanding of when they're needed. In the meantime, discussion of this area is definitely welcomed.

- While the sample AppMessage app compiles successfully (albeit with the same warnings it exhibited prior to my changes), I didn't make any modifications to it to take advantage of my changes or to kickstart the iCloud account status monitoring. I made those changes in the tvOS app that I'm working and tested the changes from my app. We can discuss what needs to be done to integrate and reasonably exercise my changes in the sample app.

- I didn't make any documentation changes. I fully expect further changes to be needed, so I didn't see the point until everything is finalized.

- Finally, I didn't change the pod version number as I wasn't sure what you'd want it changed to with this many changes.

### CODE REVIEW

**Comment by evermeer** [comment]
Wow! ... Give me some time to catch up with you. 😀

**Comment by evermeer** [comment]
Well... EVCloudKitDao has grown since I first started it on June 22 2014 (first commit, i really started the week it was announced) when everything was still beta (Swift and CloudKit). I'm only a part time iOS developer. I do a lot of C# CMS (EPiServer and Sitecore) development. Because of this improving this library goes slower than I wish. Your help like you did now is very welcome!

Since I stared EVCloudKitDao it's already refactored a couple of times. Refactoring is good. It will make the library better in the end. This change will also be substantial. Maybe it's worth a version bump to 3.0.0. But I first want to play around with it some more. You are right, It would help (not only for me) If the usage of your code would be part of the demo app. I think I also have to remove the PermissionView library. They added CloudKit support long ago in a beta version, but it never has been finished and merged to the trunk. It might be easier to just remove it.

Something else:
 
I already started to be less happy with the current eventHandler structure. For some time now I thought about restructuring this and try the approach of one of these libraries:
https://github.com/Thomvis/BrightFutures ,https://github.com/mxcl/PromiseKit or https://github.com/ReactiveX/RxSwift
My current C# project ends December 15. I hope to have 2 or 3 weeks available to make a change like this. What do you think?

In short I though doing something like this:
the .connect will return a connectionObject. You could call any of the methods on it which in it's turn will also return the same connectionObject. So you could get code like this:
```
var connectionObject = EVCloudKitData.publicDB.connect(id, predicate)
connectionObject.onCompletion { result in ... }
```
or you could do:
```
var connectionObject = EVCloudKitData.publicDB.connect(id, predicate)
    .onCompletion { result in .. }
    .onAdded {item in ..}
   .onDeleted {id in ..}
```

**Comment by tbaggett** [comment]
A couple of additional thoughts... I wanted to get the stuff in front of you for discussion, but I'm not totally satisfied that everything is ready for primetime. Please don't merge any of the changes until I complete my testing, which will happen today if I don't run into issues that require more work. I will let you know in either event.

Also, I didn't really say it in my initial comments, but please feel free to pick and choose parts or for that matter to not merge any of it at all if you have concerns. I can separate out changes for each of the features (iCloud status monitoring, per-user caching, etc.) if you only want to use some of the changes.

Yes, definitely, the sample AppMessage app and the documentation will both need to be updated if any of these additions are brought in to your project. I will be happy to help with that, but it may take a few weeks for me to do so. I really need to focus on getting my tvOS app into the app store asap. :-) The PermissionView stuff didn't seem to work that well for me, but I didn't test the sample app that much.

In regards to eventHandlers vs. Promises vs. reactive programming, etc., they all feel pretty similar to me. The last reactive style example you gave, where just the handlers you want to use are strung together, does look really clean though. 

To some extent, I think all of these approaches are just a personal preference. It would be great if a generic plugin protocol could be created that would allow any of the callback approaches to be used interchangeably. I've seen that done in C# projects to allow for choosing from different popular IoC containers like Autofac, Unity, etc., for example.

For the integration work in my tvOS app, I use the NSNotificationManager notification stuff that I added almost exclusively. I think there is only one place during app initialization where I add a one-time completion handler that removes itself when called.

For the rest of my app, I created different Observable implementations that raise a "didChange" event when property changes occur. One of those implementations is an "ObservableCloudData" class that leverages your library. It handles all the notifications from EVCloudData and EVCloudKitDao. The rest of my app code only deals with observable properties in different models. I can share that code as well if you want to see more of what I'm talking about.

I will get back to you as soon as I complete my testing. Talk with you soon!

**Comment by evermeer** [comment]
Well, for the next 2 weeks I will be quite buzzy finishing my current project. (the one that pays the bills) so continue improving until you are satisfied. I probably have time to play around with it next weekend. It took me some time to get the feeling where you want to go to. I think it looks promising. 

promises, reactive programming: You are right. Most of it is semantics. It does look clean. I think what we actually are trying to solve is the publish-subscribe design pattern don't you thing? http://everythingel.se/blog/publish-subscribe-in-swift/

**Comment by tbaggett** [comment]
Yep, agreed on the publish-subscribe pattern, the Event stuff I'm using is very similar. I continue to find minor things where Swift doesn't behave as expected, committing fixes as needed. I will try to provide a basic step-by-step of how I expected my changes to be used soon.

**Comment by tbaggett** [comment]
Quick question for you... I am continuing to test and ran into an issue. I am simulating situations where items are added on one device and removed on another. I am hitting an unhanded exception if I follow these steps:

1. Launch my tvOS app with at least one of my EVCloudData object types in existence
2. Return to main Apple TV screen but leave my app running in Xcode
3. Delete the EVCloudData object record using the iCloud dashboard in a browser
4. Return to my app on the Apple TV and BOOM goes the dynamite.

The unhandled exception is occurring in the EVCloudKitDao deleteObject method. 

    /**
     Delete an object from every data collection where it's part of
     
     - parameter recordId: The recordId of the object that will be deleted
     :return: No return value
     */
    private func deleteObject(recordId: String) {
        for (filter, _) in self.data {
            let itemID: Int? = data[filter]!.EVindexOf {item in return item.recordID.recordName == recordId}
            if itemID != nil {
                data[filter]!.removeAtIndex(itemID!)
                NSOperationQueue.mainQueue().addOperationWithBlock {
                    (self.deletedHandlers[filter]!)(recordId: recordId, dataIndex:itemID!)
                    (self.dataChangedHandlers[filter]!)()
                    self.postDataDeletedNotification(filter, recordId: recordId, dataIndex: itemID!)
                }
                dataIsUpdated(filter)
            }
        }
    }

This line of code is the culprit:

    let itemID: Int? = data[filter]!.EVindexOf {item in return item.recordID.recordName == recordId}

It blows up when the item's recordName/recordId comparison is attempted. Here's the exception output:

    2015-11-30 15:52:47.748 Marquees[1818:672274] -[__NSDictionaryM recordName]: unrecognized selector sent to instance 0x147758100
    2015-11-30 15:52:47.751 Marquees[1818:672274] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSDictionaryM recordName]: unrecognized selector sent to instance 0x147758100'
    *** First throw call stack:
    (0x1857458e8 0x184db3f80 0x18574c5fc 0x1857495a0 0x18564d6cc 0x1001e18f4 0x1001e0d40 0x1001d5da4 0x1001d6278 0x1001e55f0 0x1001f2f20 0x1001db530 0x100083010 0x1000830f8 0x18a6200b0 0x18a60ef20 0x186ceb790 0x186cebb10 0x1856fcee4 0x1856fc978 0x1856fa678 0x1856296c0 0x186b34088 0x18a3d1d64 0x1000855b0 0x1851ca974)
    libc++abi.dylib: terminating with uncaught exception of type NSException
    Message from debugger: Terminated due to signal 6

Prior to executing the problem line, I inspected the item:

    Printing description of item:
    (EVCloudKitDao.EVCloudKitDataObject) item = 0x00000001476fb3f0 {
        EVReflection.EVObject = {
            ObjectiveC.NSObject = {}
        }
        recordID = 0x0000000147758100 {
            ObjectiveC.NSObject = {}
        }
        recordType = "Performance"
        creationDate = 0x000000014775cab0 2015-11-30 21:48:51 UTC
        creatorUserRecordID = Some {
            Some = 0x0000000147757ec0 {
                ObjectiveC.NSObject = {}
            }
        }
        modificationDate = 0x000000014775cb20 2015-11-30 21:48:51 UTC
        lastModifiedUserRecordID = Some {
            Some = 0x000000014773f5e0 {
                ObjectiveC.NSObject = {}
            }
        }
        recordChangeTag = "ihmhmw90"
        encodedSystemFields = 0 bytes {
            Some = 0x000000014773f060 0 bytes
        }
    }

Have you seen this issue before? I'm not sure if any of my recent changes would have impacted this or not.

**Comment by tbaggett** [comment]
Here's some basic info on the changes that require modifications to dependent code. Other changes, like the per-user caching, "just work" without any changes required on the user's part.

**Completion Handlers**

The completion handlers for any public or private database container are added independently of  constructors or EVCloudData's "connect" method. Doing so allows handlers to be added before retrieving container instances using the static accessors. Adding handlers is done by calling the new  "add____DBInitializationCompleteHandler" methods for both the default and named containers. 

Since the db containers may be accessed via singletons from many code sources, it is unlikely that all handlers will be added before a singleton instance is initialized. I added support for recognizing newly-added handlers after an instance has already been created. Those handlers are called the next time the static instance is referenced.

I also added internal "insert____DBInitializationCompleteHandler" methods to insert handlers at the top of the list so that EVCloudData can add a default handler that closes all connections before any user-added handlers that will be (re)connecting things are called.

The DBInitializationCompleteHandlerToken instance that is returned by the "add____DBInitializationCompleteHandler" methods should be retained by the calling code. The class's releaseToken method can be used to remove the callback from the list it was assigned to.

**AppDelegate Changes**

I use completion handlers in my AppDelegate didFinishLaunchingWithOptions method as part of the initializing of EVCloudKitDao with the following:

        // Add EVCloudKit completion handlers to be called when the iCloud DB is initialized
        EVCloudKitPrivateDbToken = EVCloudData.addPrivateDBInitializationCompleteHandler(MarqueesAppModel.containerName, handler: self.fetchEVCloudKitPrivateChangeNotifications)
        EVCloudKitPublicDbToken = EVCloudData.addPublicDBInitializationCompleteHandler(MarqueesAppModel.containerName, handler: self.fetchEVCloudKitPublicChangeNotifications)
        
        // Reference the public and private DBs that we'll be using so they will be spun up.
        let _ = EVCloudData.publicDBForContainer(MarqueesAppModel.containerName)
        let _ = EVCloudData.privateDBForContainer(MarqueesAppModel.containerName)

The completion handlers are identical except for dealing with the public or private DBs.

    private func fetchEVCloudKitPrivateChangeNotifications(status: CKAccountStatus, error: NSError?) {
        
        // Release the token so we aren't called again
        EVCloudKitPrivateDbToken?.releaseToken()
        EVCloudKitPrivateDbToken = nil

        // Process any subscription notifications that were sent when we weren't running
        EVCloudData.privateDBForContainer(MarqueesAppModel.containerName).fetchChangeNotifications({
            EVLog("All private change notifications are processed")
        })
    }

I also make the new call for EVCloudKitDao to add the iCloud account status observer and a NSNotification observer of my own to watch for change notifications from EVCloudKitDao. I reinstantiate my UI in response to these notifications.

        // Add observer for iCloud account changes
        EVCloudKitDao.addiCloudAccountObserver()
        
        // Add our own observer to be notified of iCloud account changes
        NSNotificationCenter.defaultCenter().addObserverForName(EVCloudKitDao.iCloudAccountStatusChangeNotification, object: nil, queue: nil) { [unowned self] notification in
            // Reset our UI when the iCloud account info changes
            self.initializeTVControllerComponents(nil)
        }
        
Finally, I add an applicationDidBecomeActive method to call the new verifyiCloudAccountStatus method:

    func applicationDidBecomeActive(application: UIApplication) {
        // Verify iCloud account status hasn't changed
        EVCloudKitDao.verifyiCloudAccountStatus()
    }
    
**EVCloudKitDao Utilization**

I have centralized all access to EVCloudKitDao into an "ObservableCloudData" class based on the solution discussed at http://blog.scottlogic.com/2015/02/11/swift-kvo-alternatives.html. I haven't included that in my commits, but you're welcome to it if you're interested. As we've discussed, this is a bit out of scope for the EVCloudKitDao library. There are also lots of different solutions already out there for the problems this class solves for me. I'm happy with it so far, but understand that doesn't mean anyone else would be, haha. Hopefully Apple will bake something into Swift that does a great job of addressing this need.

**Comment by evermeer** [comment]
one remark: You have a static publicDBInitializationCompleteHandlers. Are you sure they do not to be an instance of the database? And if they are static, then why are you passing this on between function calls? You could just use the static instance.

**Comment by evermeer** [comment]
I created a slack channel https://evcloudkitdao.slack.com/ with github and travis integration. If you like to be added, then send me a mail to edwin@evict.nl and I will add that email address to it.

**Comment by tbaggett** [comment]
Hi Edwin, email sent. Thanks for setting that up.

Re the static HandlerCollection instances: They are static so you can use the class "add____DBInitializationCompleteHandler" methods to add your handlers before and/or independently of an instance being constructed via the publicDB, privateDB, and ...forContainer accessors. 

The intended next step was to pass the appropriate static instance to the constructor where it would be assigned to an instance property. The instance property was supposed to be used whenever the completion handlers needed to be called.

After reviewing the code to provide this response, I see that isn't how it is currently implemented. I had declared an instance property to store the reference, but never assigned the static instance to it. Later, I deleted the instance property because I saw it was unused. I should have realized at that point that something was missing. I will clean that up and commit the changes shortly.

**Comment by tbaggett** [comment]
Hold tight, more cleanup about to be committed.

**Comment by tbaggett** [comment]
Okay, so as I was saying, the appropriate static HandlerCollection instance is now passed from the static EVCloudKitDao instance accessors into the constructor, where it is assigned to the new initializationCompleteHandlers instance property. That new property is now referred to by the instance whenever it needs to call the completion handlers that are assigned to it. Hopefully that clears it up.

**Comment by evermeer** [comment]
OK, will have a look it it tomorrow. I'm currently working on the other issues (delete and multiple containers)

**Comment by tbaggett** [comment]
FYI, the changes I just committed (https://github.com/tbaggett/EVCloudKitDao/commit/a978f258e1a7f9171f21b8dea06a2f3faab624dd) should be added to your version even if you decide not to integrate my other changes. I added code to provide a "FinalResult" notification/completion handler callback if the previously-returned cached results are current and no more results are expected.

I can add these changes directly to your version if needed once you decide about the other changes, if you like.

**Comment by tbaggett** [comment]
I would like to close this pull request if you have no plans to integrate my changes. No hard feelings, but I have been maintaining sync between our branches for two months now. It is getting harder to keep my branch in sync with yours as the number of differences increase. I will look into moving my modifications to one or more subclasses of your framework classes as needed.

**Comment by evermeer** [comment]
You are right, it's time for me to really pick this one up. I already told you I would cherry pick from this pull request for more than a month. By now I still have not made up my mind how to handle this thing. Since I am planning to cherry pick and not merge, it's definitely ok to close this pull request.
